lib/spack/llnl/util/symlink.py

```diff
@@ -9,59 +9,136 @@
 from os.path import exists, join
 from sys import platform as _platform
 
+import llnl.util.tty as console
 from llnl.util import lang
 
 is_windows = _platform == "win32"
 
 if is_windows:
+    import subprocess
+
     from win32file import CreateHardLink
 
 
 def symlink(real_path, link_path):
     """
-    Create a symbolic link.
-
-    On Windows, use junctions if os.symlink fails.
+    Try to create a symbolic link.
+
+    On non-Windows and Windows with System Administrator
+    privleges this will be a normal symbolic link via
+    os.symlink.
+
+    On Windows without privledges the link will be a
+    junction for a directory and a hardlink for a file.
+    On Windows the various link types are:
+
+    Symbolic Link: A link to a file or directory on the
+    same or different volume (drive letter) or even to
+    a remote file or directory (using UNC in its path).
+    Need System Administrator privileges to make these.
+
+    Hard Link: A link to a file on the same volume (drive
+    letter) only. Every file (file's data) has at least 1
+    hard link (file's name). But when this method creates
+    a new hard link there will be 2. Deleting all hard
+    links effectively deletes the file. Don't need System
+    Administrator privileges.
+
+    Junction (sometimes called soft link): A link to a
```

Comment:
Thanks for adding that reference. I think it would be good to avoid using the term "soft link" in this context (given that we handle symlinks for Windows and Linux), even though it is correct according to the source you linked (or you could explain that "soft link" means symlink on Linux, which is different, but I'm not sure if it's worth explaining since having both terms seems of limited use).

Reply:
Agreed that it doesn't help the user that much, so removed the parenthetical.